rebalancing/rebalancing.py

```diff
@@ -0,0 +1,311 @@
+
+        while z < y:
+            y = z
+            z = ((x // z) + z) // 2
+
+        return y
+
+    def _calculate_tokens_to_retire(self, dex_score: dexTokenInterface, flag: int) -> int:
+        """
+        :param dex_score: Interface of dex score.
+        Returns the amount of sICX required for rebalancing the price.
+        """
+        oracle_score = self.create_interface_score(self._oracle.get(), oracleTokenInterface)
+        oracle_price = oracle_score.get_reference_data("USD", "ICX")
+        oracle_rate = oracle_price["rate"]
+        pool_stats = dex_score.getPoolStats(2)
+        sicx_supply = pool_stats['base']
+        bnusd_supply = pool_stats['quote']
+        value = (self._sqrt(oracle_rate * sicx_supply * bnusd_supply) // 10 ** 9) - sicx_supply
+        if flag == 1:
+            # value = (self._sqrt(oracle_rate * sicx_supply * bnusd_supply) // 10 ** 9) - bnusd_supply
+            value = (self._sqrt((sicx_supply * bnusd_supply) // oracle_rate) * 10 ** 9) - bnusd_supply
+        return value
+
+    @external
+    @only_governance
+    def setPriceChangeThreshold(self, _value: int) -> None:
+        """
+        :param _value: threshold to set.
+        Sets the threshold .
+        """
+        self._price_threshold.set(_value)
+
+    @external(readonly=True)
+    def getPriceChangeThreshold(self) -> int:
+        """
+        Returns the threshold value set by Governance contract.
+        """
+        return self._price_threshold.get()
+
+    @external
+    @only_governance
+    def setSicxReceivable(self, _value: int) -> None:
+        """
+        :param _value: sICX amount to set.
+        Sets the sICX amount to receive by rebalancing contract.
+        """
+        self._sicx_receivable.set(_value)
+
+    @external
+    @only_governance
+    def setBnusdReceivable(self, _value: int) -> None:
+        """
+        :param _value: bnUSD amount to set.
+        Sets the bnUSD amount to receive by rebalancing contract.
+        """
+        self._bnusd_receivable.set(_value)
+
+    @external(readonly=True)
+    def getSicxReceivable(self) -> int:
+        """
+        Returns the sICX amount to receive by rebalancing contract.
+        """
+        return self._sicx_receivable.get()
+
+    @external(readonly=True)
+    def getBnusdReceivable(self) -> int:
+        """
+        Returns the bnUSD amount to receive by rebalancing contract.
+        """
+        return self._bnusd_receivable.get()
+
+    @external(readonly=True)
+    def getRebalancingStatus(self) -> list:
+        """
+        Checks the Rebalancing status of the pool.
+        """
+        rebalancing_direction = 0
+        self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
+        self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
+        self.loans_score = self.create_interface_score(self._loans.get(), loansTokenInterface)
+        price = self.bnUSD_score.lastPriceInLoop()
```

Comment:
This is the same as the oracle price on line 194.
getPoolStats has everything you need to get the price on line 262, so you can save calls to the oracle and the DEX. Get the number of tokens in the pool here with getPoolStats, get the oracle price, pass them to the _calculate_tokens_to_retire method.
Also, you have missed the price of sICX in loop, which you need to get the true price to compare with the DEX price.

You do this calculation three times - `10 ** 36 // pool_price_dex`
Assign `dex_price = pool_stats['base'] * EXA // pool_stats['quote']`
    where, `EXA = 10 ** 18`

Try defining a constant `TOKENS = ["sICX", "bnUSD"]`
then,
```
def getRebalancingStatus(self) -> (bool, int, str):
    self.bnUSD_score = self.create_interface_score(self._bnUSD.get(), bnUSDTokenInterface)
    self.dex_score = self.create_interface_score(self._dex.get(), dexTokenInterface)
    self.sICX_score = self.create_interface_score(self._sicx.get(), sICXTokenInterface)
    price = self.bnUSD_score.lastPriceInLoop() * EXA // self.sICX_score.lastPriceInLoop()
    pool_stats = dex_score.getPoolStats(2)
    dex_price = pool_stats['base'] * EXA // pool_stats['quote']
    direction = price < dex_price
    diff = (dex_price - price) * EXA // price
    max_diff = self._price_threshold.get()
    if not -max_diff < diff < max_diff:
        return True, self._calculate_tokens_to_retire(price, pool_stats['base'], pool_stats['quote'], direction), TOKENS[direction]
    return False, 0, ""
```